src/main/scala/org/graphframes/Pregel.scala

```diff
@@ -0,0 +1,234 @@
+    this
+  }
+
+  /**
+   * Set the message column. In each round of pregel, each triplet
+   * (src-edge-dst) will generate zero or one message and the message will
+   * be sent to the src vertex of this triplet.
+   *
+   * @param msgExpr The message expression. It is a sql expression and it
+   *                can reference all propertis in the triplet, in the way
+   *                `Pregel.src("src_col_name)`, `Pregel.edge("edge_col_name)`,
+   *                `Pregel.dst("dst_col_name)`
+   */
+  def sendMsgToSrc(msgExpr: Column): this.type = {
+    sendMsgs += Tuple2(Pregel.src("id"), msgExpr)
+    this
+  }
+
+
+  /**
+   * Set the message column. In each round of pregel, each triplet
+   * (src-edge-dst) will generate zero or one message and the message will
+   * be sent to the dst vertex of this triplet.
+   *
+   * @param msgExpr The message expression. It is a sql expression and it
+   *                can reference all propertis in the triplet, in the way
+   *                `Pregel.src("src_col_name)`, `Pregel.edge("edge_col_name)`,
+   *                `Pregel.dst("dst_col_name)`
+   */
+  def sendMsgToDst(msgExpr: Column): this.type = {
+    sendMsgs += Tuple2(Pregel.dst("id"), msgExpr)
+    this
+  }
+
+  /**
+   * Set the aggregation expression which used to aggregate messages which received
+   * by each vertex.
+   *
+   * @param aggExpr The aggregation expression, such as `sum(Pregel.msgCol)`
+   */
+  def aggMsgs(aggExpr: Column): this.type = {
+    aggMsgsCol = aggExpr
+    this
+  }
+
+  /**
+   * After set a series of things via above methods, then call this method to run
+   * pregel, and it will return the result vertex dataframe, which will include all
+   * updated columns in the final rounds of pregel.
+   *
+   * @return the result vertex dataframe
+   */
+  def run(): DataFrame = {
+    val sendMsgsColList = sendMsgs.toList.map { case (id, msg) =>
+      struct(id.as("id"), msg.as("msg"))
+    }
+
+    val initVertexCols = withVertexColumnList.toList.map { case (colName, initExpr, _) =>
+      initExpr.as(colName)
+    }
+    val updateVertexCols = withVertexColumnList.toList.map { case (colName, _, updateExpr) =>
+      when(Pregel.msg.isNotNull, updateExpr).otherwise(col(colName)).as(colName)
+    }
+
+    var vertexUpdateColDF = graph.vertices.select((col("id") :: initVertexCols): _*)
+    var vertices = graph.vertices.join(vertexUpdateColDF, "id")
+    val edges = graph.edges
+
+    var iteration = 1
+
+    val shouldCheckpoint = checkpointInterval > 0
+
+    while (iteration <= maxIter) {
+      val beginTime = System.currentTimeMillis()
+
+      val tripletsDF = vertices.select(struct(col("*")).as("src"))
+        .join(edges.select(struct(col("*")).as("edge")), Pregel.src("id") === Pregel.edge("src"))
+        .join(vertices.select(struct(col("*")).as("dst")), Pregel.edge("dst") === Pregel.dst("id"))
+
+      assert(sendMsgs.length > 0)
```

Comment:
Does this need to be in the loop? Can we raise a better error message if this is false.